src/Features/Core/Portable/Diagnostics/DefaultDiagnosticAnalyzerService.cs

```diff
@@ -78,53 +80,73 @@ public bool NeedsReanalysisOnOptionChanged(object sender, OptionChangedEventArgs
             {
+                Debug.Assert(document.Project.Solution.Workspace == _workspace);
+
                 // right now, there is no way to observe diagnostics for closed file.
                 if (!_workspace.IsDocumentOpen(document.Id) ||
-                    !_workspace.Options.GetOption(InternalRuntimeDiagnosticOptions.Syntax) ||
-                    !document.SupportsSyntaxTree)
+                    !_workspace.Options.GetOption(InternalRuntimeDiagnosticOptions.Syntax))
                 {
                     return;
                 }
 
-                var tree = await document.GetSyntaxTreeAsync(cancellationToken).ConfigureAwait(false);
-                var diagnostics = tree.GetDiagnostics(cancellationToken);
-
-                Debug.Assert(document.Project.Solution.Workspace == _workspace);
-
-                var diagnosticData = diagnostics == null ? ImmutableArray<DiagnosticData>.Empty : diagnostics.Select(d => DiagnosticData.Create(document, d)).ToImmutableArrayOrEmpty();
-
-                _service.RaiseDiagnosticsUpdated(
-                    DiagnosticsUpdatedArgs.DiagnosticsCreated(new DefaultUpdateArgsId(_workspace.Kind, Syntax, document.Id),
-                    _workspace, document.Project.Solution, document.Project.Id, document.Id, diagnosticData));
+                await AnalyzeForKind(document, AnalysisKind.Syntax, cancellationToken).ConfigureAwait(false);
             }
 
             public async Task AnalyzeDocumentAsync(Document document, SyntaxNode bodyOpt, InvocationReasons reasons, CancellationToken cancellationToken)
             {
-                // right now, there is no way to observe diagnostics for closed file.
-                if (!_workspace.IsDocumentOpen(document.Id) ||
-                    !_workspace.Options.GetOption(InternalRuntimeDiagnosticOptions.Semantic))
+                Debug.Assert(document.Project.Solution.Workspace == _workspace);
+
+                if (!IsSemanticAnalysisOn())
                 {
                     return;
                 }
 
-                var model = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
-                var diagnostics = model.GetMethodBodyDiagnostics(span: null, cancellationToken: cancellationToken).Concat(
-                                    model.GetDeclarationDiagnostics(span: null, cancellationToken: cancellationToken));
+                await AnalyzeForKind(document, AnalysisKind.Semantic, cancellationToken).ConfigureAwait(false);
 
-                Debug.Assert(document.Project.Solution.Workspace == _workspace);
+                bool IsSemanticAnalysisOn()
+                {
+                    // right now, there is no way to observe diagnostics for closed file.
+                    if (!_workspace.IsDocumentOpen(document.Id))
+                    {
+                        return false;
+                    }
```

Comment:
I'm not sure i understand this.  Why is semantic analysis not on for closed files?  Isn't that what FSA is?

Reply:
can this be answered?